java/src/main/native/src/row_conversion.cu

```diff
@@ -1238,6 +1241,56 @@ copy_validity_from_rows(const size_type num_rows, const size_type num_columns,
 #endif // ASYNC_MEMCPY_SUPPORTED
 }
 
+/**
+ * @brief copies string data from jcudf row format to cudf columns
+ *
+ * @tparam RowOffsetIter iterator for row offsets into the destination data
+ * @param row_offsets offsets for each row in input data
+ * @param string_row_offsets offset data into jcudf row data for each string
+ * @param string_lengths length of each incoming string in each column
+ * @param string_column_offsets offset column data for cudf column
+ * @param string_col_data output cudf string column data
+ * @param row_data jcudf row data
+ * @param num_rows number of rows in data
+ * @param num_string_columns number of string columns in the table
+ */
+template <typename RowOffsetIter>
+__global__ void copy_strings_from_rows(RowOffsetIter row_offsets, int32_t **string_row_offsets,
+                                       int32_t **string_lengths, size_type **string_column_offsets,
+                                       char **string_col_data, int8_t const *row_data,
+                                       size_type const num_rows,
+                                       size_type const num_string_columns) {
+  // each warp takes an element striding on grid dimension. It will copy that string to the final
+  // destination. Traversing in row-major order to coalesce the offsets and size reads.
+  auto my_block = cooperative_groups::this_thread_block();
+  auto my_partition = cooperative_groups::tiled_partition<32>(my_block);
+#ifdef ASYNC_MEMCPY_SUPPORTED
+  cuda::barrier<cuda::thread_scope_block> block_barrier;
+#endif
+
+  // workaround for not being able to take a reference to a constexpr host variable
+  auto const ROWS_PER_BLOCK = NUM_STRING_ROWS_PER_BLOCK_FROM_ROWS;
+  auto const tiles_per_col = util::div_rounding_up_unsafe(num_rows, ROWS_PER_BLOCK);
+  auto const starting_tile =
+      blockIdx.x * my_partition.meta_group_size() + my_partition.meta_group_rank();
+  auto const num_tiles = tiles_per_col * num_string_columns;
+  auto const tile_stride = my_partition.meta_group_size() * gridDim.x;
+  for (auto my_tile = starting_tile; my_tile < num_tiles; my_tile += tile_stride) {
+    auto const starting_row = (my_tile % tiles_per_col) * ROWS_PER_BLOCK;
+    auto const col = my_tile / tiles_per_col;
+    auto const str_len = string_lengths[col];
+    auto const str_row_off = string_row_offsets[col];
+    auto const str_col_off = string_column_offsets[col];
+    auto str_col_data = string_col_data[col];
+    for (int row = starting_row; row < starting_row + ROWS_PER_BLOCK && row < num_rows; ++row) {
```

Comment:
This actually was happening here. It was well hidden by the `MEMCPY` macro. The interface to have multiple threads participate requires passing the cooperative group as the first parameter. I've updated the code to do that and verified the output with tests. Thanks for bringing this up because I had to think about it some more and looked up the `memcpy_async` API again to ensure I had done it correctly.

As for the comments, I've added one here describing this loop and updated the comment at the top of the function talking about the overall function. If you'd like more, please let me know.